### DISCUSSION TITLE
A hole in the ground.

### DISCUSSION THREAD

**Discussion by dthery** [google/model-viewer#2204] [Q&A]
### Description

Can we consider placing a model under the ground?
With the example below, I would like the cylinder to be in the ground and the cube on the ground.
We should be able to look inside as if it were a well.

#### Live Demo

https://www.dithernet.com/trou
Download the model here:
https://www.dithernet.com/Astro/etalon.glb

**Comment by elalish** [comment]
That's interesting. Perhaps we could introduce an attribute like `ar-y-offset`? That would allow for holes and also floating objects. How does that sound?

**Comment by ronpadz** [comment]
A few thoughts on this:
- What would the shadow behavior be?
- Seems to be related to #998

**Comment by dthery** [comment]
Yes @elalish,  it would be interesting. Thank's.

**Comment by elalish** [comment]
@ronpadz Thanks for linking to #998; something like `normal-offset` might be more generic for also handling placement on vertical surfaces as in #687.

I think the shadow should stay on the plane the object is offset from. Perhaps it should even expand the bounding box for the purposes of auto framing? When we get into vertical placement there's also the question of whether the shadow should be on the floor or the wall.

**Comment by dthery** [comment]
Hi, Any news for this opportunity?

**Comment by elalish** [comment]
@dthery It's not on the short list, so I can't promise when we'll get to it. Would you be interested in contributing?

**Comment by dthery** [comment]
@elalish I'm willing to help, if it's within my skill set :-)

**Comment by elalish** [comment]
@dthery Fantastic! I think a better attribute name would be `surface-offset` and you could add that to ar.ts. Then pass it in to the ARRenderer and translate the model accordingly. Open a PR when you're ready for feedback or help.

**Comment by dromeuf** [comment]
Hello, I am looking to represent archaeological excavations in AR. With ARCore, I can't place the 3D model under the ground but just above it. Would there be a way or a trick to shift the 3D model in the ground, without even taking into account the intersections with real objects that would be above the ground, and therefore logically above the model. It would simply require a filling of the area in the ground of the 3D model for my archaeological application. Finally, thank you for your work !

**Comment by elalish** [comment]
@dromeuf That's exactly the feature this issue is discussing. Perhaps it's time to prioritize it. As for a hack in the meanwhile, the only thing I can think of is to make use of glTF animations. We calculate the bounding box (the bottom of which we place on the floor) based on the resting state. An animation could move it down and we would not compensate. Perhaps you can make an animation that doesn't actually move, but just provides this static offset.

**Comment by dromeuf** [comment]
ah yes, it would be great if you could treat objects and areas in the ground with ARCore. I have several thousand square meters that our university students have been excavating archaeologically for 20 years! They would be happy to see their work valued by a modern media. These are Iron Age excavations, and from the Roman period in Gaul (Corent France). With all our thanks in advance. If you need a 3D model as an example, I can transfer one to you in a private message.

**Comment by dthery** [comment]
Cool, I'll try the hack you're talking about.

**Comment by dthery** [comment]
Super @elalish  your hack works. However, the first key of the animation frame must be at 0 on the vertical axis and then you push the object into the next key and stop the animation.

**Comment by dthery** [comment]
another problem is added for this effect. It should be possible to apply a material with the texture of the camera image. This would allow to hide the part under the ground normally invisible.  On Unity I do it with a special shader. Do you have a trick? https://imgur.com/wPFQzjz  I would like to hide the red zone.

**Comment by elalish** [comment]
No, we don't support custom shaders. That is an interesting problem and I can't think of a good solution at the moment.

**Comment by klaasnienhuis** [comment]
I'd also be very interested in a y-offset feature. I'm working with swimming pools which I'd like to put in the ground instead of on top of it.

@dthery could you tell me how you've made the animation? I've animated my pool to move downwards. On key 0 it's on the base height, on key 1 it's pushed down. However, in AR on my android device, the pool is still placed on top of the scanned floor. I can confirm the object has been successfully animated, though the AR process seems to look at the boundingbox at the current frame, not only frame 0.
On ios, with a USDZ-version of the GLB file I used on android, the animation trick does work. The pool is placed with the boundingbox from frame 0 and then the pool is animated downwards on frame 1 and stays there.

@elalish could you comment on this, or confirm whether on android the boundingbox of the current animation frame is used for model placement?

**Comment by elalish** [comment]
Are you using scene-viewer or webxr mode? My method is likely only to work in WebXR mode, as that's the one I control.

**Comment by dthery** [comment]
@klaasnienhuis I did exactly as you said and it works for me. But I haven't found a solution to hide the areas that shouldn't be visible in the ground. So the effect is not realistic.
